### ISSUE TITLE
Question: multiple cameras/views onto the same scene

### ISSUE THREAD

**Issue by os-gabe** [vispy/vispy#1666]
I am creating a Qt app for viewing some 3D elements and would like to create just one scene then display multiple interactive views into that scene.

Currently I have something like this:
```
class MyCanvas(scene.SceneCanvas):
    def __init__(self, parent):
        scene.SceneCanvas.__init__(self)
        self.unfreeze()
        self.parent = parent

        self.view = self.central_widget.add_view()
        self.view.camera = "arcball"

        self.visual = scene.visuals.Markers()

        self.view.add(self.visual)
        self.freeze()

class MyWidget(QtWidgets.QWidget):
    def __init__(self, parent):
        QtWidgets.QWidget.__init__(self)

        self.canvas = MyCanvas(parent=parent)
        self.canvas.create_native()
        self.canvas.native.setParent(self)

        self.layout = QtWidgets.QVBoxLayout(self)
        self.layout.addWidget(self.canvas.native)
```
I want to add another view that uses a panzoom camera then display both of these in my Qt app. Any suggestions or examples you can point me to?

**Comment by djhoese** [comment]
I would start with the `viewbox` example: https://github.com/vispy/vispy/blob/master/examples/basics/scene/viewbox.py

This (iirc) creates two views with the same data. It does not connect the two views but I believe there are other examples that do things like this if you're interested. Let me know if this doesn't answer your question.

**Comment by os-gabe** [comment]
Thanks for the response @djhoese. After looking through the examples a bit more I realized what I actually want is multiple canvases with visuals shared between them. The example here does exactly that:
https://github.com/vispy/vispy/blob/master/examples/basics/scene/shared_context.py

Unfortunately it doesn't work with PyQt5 so maybe what I want is not possible at the moment. I can always just make two independent canvases and manage keeping the visuals in sync between them manually - that's just more work.

**Comment by djhoese** [comment]
So shared contexts are meant to work, but the original version of pyqt5 I had access to from conda-forge was buggy and I couldn't get it to work. There is now a newer version of pyqt that you could try. The shared contexts can be enabled with an environment variable if you check the qt backend source code. I'm traveling back from SciPy 2019 right now so can't look it up easily. Sorry.

**Comment by os-gabe** [comment]
I see. Adding `os.environ["VISPY_PYQT5_SHARE_CONTEXT"] = "true"` to the top of the `shared_context.py` example makes it run.

However, the behavior isn't as I expected. A visual is created and added to two different canvases (where the second canvas shares context with the first). The visual only shows up in whichever canvas you add it to last though. This is no different than the case where you do not share context.

 I'm not sure if this is a bug or if I'm misunderstanding the expected behavior.

**Comment by os-gabe** [comment]
Tested with `PyQt5-5.12.1` and `PyQt5-5.13.0`

**Comment by djhoese** [comment]
What operating system are you on? Where did you get the pyqt from? I'll see what I can get with 5.9 from conda-forge.

**Comment by os-gabe** [comment]
I'm running Ubuntu 18.04 and pip installing PyQt from https://pypi.org/project/PyQt5/

**Comment by djhoese** [comment]
Based on a comment from @larsoner in #862 I think shared contexts may not work at the SceneGraph level. Oddly our two shared_context.py examples both use the SceneCanvas.

The main issue seems to be a change a while ago that stopped visuals from allowing multiple parents. So when we set the parent for the SurfacePlot visual the second time it causes a complete disconnection for that visual from the first view/Scene. I can try to throw together a gloo-based example, but to use cameras and scenes will require some major changes in VisPy.

@larsoner @kmuehlbauer Does this sound completely unreasonable to support in the Scene? The GL program/visual has to be explicitly drawn in each context/view that it is used in so besides handling multiple parents carefully, is this functionally impossible?

**Comment by djhoese** [comment]
Oh, wouldn't transforms completely break? There would have to be a constraint that the multi-parents have to all be viewboxes...I think.

**Comment by larsoner** [comment]
I don't recall what the problem was. From what I recall, the complexity was a bit over my head at the time to fully understand and I definitely don't at this point. Some GitHub sleuthing could probably show some discussion of it. If you can't find it easily let me know and I'll look

**Comment by djhoese** [comment]
So the main conversation seems to be in #402 and mentions the huge PR #1021 where multi-parents were removed. @almarklein in #402 brings up exactly what I thought, transforms would conflict between two cameras (or even if the user changes a parent transform for one canvas, but not the other).

**Comment by kmuehlbauer** [comment]
@os-gabe @larsoner @djhoese see #1124 for comments from @campagnola on how to tackle this.

**Comment by almarklein** [comment]
FWIW in another (web-based, non-oss) 3D viz project we've got this working in a pretty nice way. It would take a huge refactoring to adapt such an approach in vispy, but I thought I'd share it as "background info".

In that project, we decided on not giving visuals a parent property. Each visual keeps track of its children, and a visual could be part of multiple "parent" visuals. By not having an explicit parent (or parent**s**) property, the code stays relatively simple. The chain of visuals is passed to the draw event, in case the visual needs it. The only thing is that the visual must expect to be drawn multiple times, with different chains of parents.

**Comment by djhoese** [comment]
I've been thinking about this a lot and going over @campagnola's comment [here](https://github.com/vispy/vispy/issues/1124#issuecomment-245499165) and I think I'm finally starting to get a handle on things. In summary, VisPy's SceneCanvas, ViewBox, BaseCamera, and TransformSystem are incredibly complex.

Flipping the parent/children relationship might simplify the user's experience of "this visual belongs to Scene X and Scene Y", but it doesn't solve the other issues with multiple parents that are in VisPy. After re-reviewing the code the main issue is the transform system and Node classes access to the root canvas and various properties of it. For example, all TransformSystem's `dpi` properties access the canvases `dpi` property. All `TransformSystem` objects have handles on the transforms of various important nodes in the graph so it can easily generate chains of transforms from node X to node Y. I've always disliked the need for a `Node` class but now that I see how much work it is actually doing (managing transforms, etc) I can see that it is no small/simple monster.

I was hoping that this discussion would lead to simplifying the entire SceneCanvas and TransformSystem but at this point I can't pretend that any refactor won't end up with the same end result. One great thing would be to make transforms more "static" so they don't need to get recreated all the time (see `ChainTransform` usage `node.py`). This would help reduce any recompiling of the vertex shader code and all the perform issues that can result from that.

Bottom line, I think @campagnola's suggestions are still our best bet and getting this to work.
